287 moving columns to multiindex

ktroutman opened the pull request:
Here is my initial shot at the re/set index functionality. I tried to stay tight on the pandas implementation. Would be interested in thoughts or suggestions.

cosmicBboy commented on pandera/schemas.py:
```diff
@@ -689,6 +690,181 @@ def to_yaml(self, fp: Union[str, Path] = None):
+        #
+        additional_columns: list = []
+        new_index = new_schema.index
+        if level is None:
```
the conditional branching from lines 817-861 makes the logic a little bit difficult to follow. Consider an alternative approach, something a little flatter like:

```python
        # make sure level is a consistent type, similar to lines 803-804, but
        # now level is guaranteed to be a list
        level = (
            [] if level is None
            else [level] if not isinstance(level, list)
            else level
        )

        # get names of all indexes
        index_levels = (
            [index.name for index in new_schema.index.indexes]
            if isinstance(new_schema.index, MultiIndex)
            else [new_schema.index.name]
        )

        # determine additional columns
        additional_columns = {} if drop else {
            key: new_schema.index.columns[key]
            for key in index_levels if key in level
        }

        # determine new index
        new_index: Optional[Union[Index, MultiIndex]]
        if isinstance(new_schema.index, MultiIndex):
            new_index = MultiIndex([
                index for index in new_schema.index.indexes
                if index.name not in level
            ])
        elif new_schema.index.name not in level:
            new_index = new_schema.index
        else:
            new_index = None

        new_schema = new_schema.add_columns({
            col: new_schema.index.columns[col] for col in additional_columns
        })
        new_schema.index = new_index
```

cosmicBboy reviewed (commented):
nice work @ktroutman!

As a procedural thing, I think it would help the dev process if you [set up pre-commit](https://github.com/pandera-dev/pandera/blob/master/.github/CONTRIBUTING.md#set-up-pre-commit), which handles code formatting and linting before pushing to remote, as well as `make tests` to make sure all tests are passing (I saw the circular import error when I ran this PR locally, see inline comments).

Overall I think the behavior is what we want, my main thought about the API is that I think we can go without the `inplace` argument. The main reason it's there in `pandas` is for memory efficiency and avoiding copies of potentially large dataframes. In `pandera`'s case, however, the memory footprint of schemas are negligible, so I think it would be desirable to always copy schemas when they're modified to prevent unexpected schema mutation (and it also simplifies the implementation!).

I added some more detailed feedback in the inline comments, feel free to ignore the `nit` comments!

ktroutman commented:
Thanks a lot for the comments, I was struggling a bit on getting the tests to run (for technical reasons) and the hooks to run. I've hacked through the technical issues on my side and I'm pretty sure I've addressed most of your comments in the latest commit.

I found the Multiindex and Index stuff to be cumbersome to work with (if you remove one column from a 2 column multiindex, must then manually change to Index), with this feature as well as in working with the api on my projects. If this idea does work, I would probably use this as my main interface with indices, similar to pandas.

cosmicBboy commented on pandera/schemas.py:
```diff
@@ -6,22 +6,22 @@
 from functools import wraps
 from pathlib import Path
-from typing import Any, Callable, Dict, List, Optional, Union
+from typing import Any, Callable, Dict, List, Optional, Tuple, Union
```
`Tuple` is unused

cosmicBboy commented on pandera/schemas.py:
```diff
@@ -6,22 +6,22 @@
-from .checks import Check
-from .dtypes import PandasDtype, PandasExtensionType
-from .error_formatters import (
+from pandera import constants, dtypes, errors
```
these should be reverted to relative imports

ktroutman replied:
sure thing -- this is a remnant of my environment set up issues. will revert on next commit.

cosmicBboy commented on pandera/schemas.py:
```diff
@@ -689,6 +689,168 @@ def to_yaml(self, fp: Union[str, Path] = None):
+            ]
+            assert not_in_cols == []
+        except AssertionError:
+            raise Exception(f"Keys {not_in_cols} not found in schema columns!")
```
these exceptions should be `errors.SchemaInitError` and use the `from` keyword so that full error trace is available. `assert` statements are also nice when quickly scripting things, but in this case it might be more appropriate to use a conditional then `raise` here.

```python
not_in_cols: List[str] = [
    x for x in keys_temp if x not in new_schema.columns.keys()
]
if not not_in_cols:
     raise errors.SchemaInitError(f"Keys {not_in_cols} not found in schema columns!")
```

cosmicBboy commented on pandera/schemas.py:
```diff
@@ -689,6 +689,168 @@ def to_yaml(self, fp: Union[str, Path] = None):
+            dup_cols: List[str] = [x for x in set(keys_temp) if keys_temp.count(x) > 1]
+            assert dup_cols == []
+        except AssertionError:
+            raise Exception(f"Keys {dup_cols} are duplicated!")
```
see comment above

cosmicBboy commented on pandera/schemas.py:
```diff
@@ -689,6 +689,168 @@ def to_yaml(self, fp: Union[str, Path] = None):
+        try:
+            assert new_schema.index is not None
+        except AssertionError:
+            raise Exception("There is currently no index set for this schema.")
```
see comment on line 717

cosmicBboy commented on pandera/schemas.py:
```diff
@@ -689,6 +689,168 @@ def to_yaml(self, fp: Union[str, Path] = None):
+        )
+        try:
+            assert level_not_in_index == []
+        except AssertionError:
```
see comment on line 717

cosmicBboy commented on pandera/schemas.py:
```diff
@@ -689,6 +689,168 @@ def to_yaml(self, fp: Union[str, Path] = None):
+        level_temp: Union[List[Any], List[str], None] = list(set(level)) if level is not None else []
+
+        # ensure all specified keys are present in the index
+        level_not_in_index: List[str] = (
```
the type of this variable should be compatible with `level_temp`, since it's a potential type of this value.

cosmicBboy commented on pandera/schemas.py:
```diff
@@ -689,6 +689,168 @@ def to_yaml(self, fp: Union[str, Path] = None):
+
+
+        # ensure no duplicates and tuple type
+        level_temp: Union[List[Any], List[str], None] = list(set(level)) if level is not None else []
```
I don't think `None` should be an allowable type type, since this'll be an empty list of `level is None`

cosmicBboy reviewed (commented):
it's looking good @ktroutman! looks like unit tests are passing, a few more things to ironed out in code-formatting land, looks `schemas.py` and `test_schemas.py` need to re-formatted with `black` (needs to be version `20.8b1` with option `--line-length 79` if pre-commit is finicky and you need to do it manually)

And then some [pylint errors](https://travis-ci.org/github/pandera-dev/pandera/jobs/744045589#L695-L716), some with code formatting issues. See my in-line comments for specifics on the `raise-missing-from` error.

Also, let me know if you need any help setting up a development environment with all the testing infrastructure on it, I realize the developer docs are a little sparse and would be worth making it a little explicit!

ktroutman commented:
Thanks again for the quick response and really helpful comments! I've fixed main issues youve pointed out and the type checking seems to be ok now.

As far as testing goes, I tested the main functionality that I can think of, but something that could I could probably use advice on :)

cosmicBboy commented:
thanks @ktroutman, the tests look good to me! I pushed a few changes to your branch to get pylint and black to stop complaining

ktroutman commented on pandera/schemas.py:
```diff
@@ -699,14 +699,16 @@ def set_index(
 
         keys_temp: List = (
-            list(set(keys)) if not isinstance(keys, List) else keys
+            list(set(keys)) if not isinstance(keys, list) else keys
```
💡
